Zend/zend_object_handlers.c

```diff
@@ -728,6 +784,13 @@ ZEND_API zval *zend_std_write_property(zend_object *zobj, zend_string *name, zva
 			Z_TRY_ADDREF_P(value);
 
 			if (UNEXPECTED(prop_info)) {
+				if (UNEXPECTED(prop_info->flags & ZEND_ACC_READONLY)) {
```

Comment:
I assume the value cannot be initialized in a subclass on a redeclared property if the parent constructor is called and the property is already initialized?

```php
class A {
    public readonly int $prop;
    
    public function __construct() {
        $this->prop = 1;
    }
}
class B extends A {
    public readonly int $prop;

    public function __construct() {
        parent::__construct();
        $this->prop = 2;
    }
}
```

This could be solved by initializing the property in a separate method.

```php
class A {
    public readonly int $prop;
    
    public function __construct() {
        $this->initializeProp();
    }

    protected function initializeProp() {
        $this->prop = 1;
    }
}
class B extends A {
    public readonly int $prop;

    protected function initializeProp() {
        $this->prop = 2;
    }
}
```

Or computing the value in a separate method.

```php
class A {
    public readonly int $prop;
    
    public function __construct() {
        $this->prop = $this->computeProp();
    }

    protected function computeProp() {
        return 1;
    }
}
class B extends A {
    protected function computeProp() {
        return 2;
    }
}
```

In this case the property doesn't need to be overridden. I wonder if solution 2 is similar enough that allowing the property to be redefined might not even be necessary?

Reply:
I don't think redeclaring a property is ever useful right now. The question is more whether it makes sense to forbid this for readonly properties while it is allowed for other properties.

Reply:
I forgot that you might want to change the visibility in the child class. So I suppose this is necessary.